src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/AADFedAuthTokenRefreshTest/AADFedAuthTokenRefreshTest.cs

```diff
@@ -0,0 +1,133 @@
+                    // Display new expiry in local time
+                    DateTime newLocalExpiryTime = TimeZoneInfo.ConvertTimeFromUtc((DateTime)newExpiry, TimeZoneInfo.Local);
+                    LogInfo($"Token: {tokenHash2}   New Expiry: {newLocalExpiryTime}");
+
+                    Assert.True(tokenHash1 == tokenHash2, "The token's hash before and after token refresh must be identical.");
+                    Assert.True(newLocalExpiryTime > oldLocalExpiryTime, "The refreshed token must have a new or later expiry time.");
+                }
+            }
+        }
+
+        private void LogInfo(string message)
+        {
+            _testOutputHelper.WriteLine(message);
+        }
+
+        private DateTime? GetOrSetTokenExpiryDateTime(SqlConnection connection, bool setExpiry, out string tokenHash)
+        {
+            try
+            {
+                // Get the inner connection
+                object innerConnectionObj = connection.GetType().GetProperty("InnerConnection", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(connection);
+
+                // Get the db connection pool
+                object poolObj = innerConnectionObj.GetType().GetProperty("Pool", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(innerConnectionObj);
+
+                // Get the Authentication Contexts
+                IEnumerable authContextCollection = (IEnumerable)poolObj.GetType().GetProperty("AuthenticationContexts", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(poolObj, null);
+
+                // Get the first authentication context
+                object authContextObj = authContextCollection.Cast<object>().FirstOrDefault();
+
+                // Get the token object from the authentication context
+                object tokenObj = authContextObj.GetType().GetProperty("Value").GetValue(authContextObj, null);
+
+                DateTime expiry = (DateTime)tokenObj.GetType().GetProperty("ExpirationTime", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(tokenObj, null);
+
+                if (setExpiry)
+                {
+                    // Forcing 1 minute expiry to trigger token refresh.
+                    expiry = DateTime.UtcNow.AddMinutes(1);
+
+                    // Apply the expiry to the token object
+                    FieldInfo expirationTime = tokenObj.GetType().GetField("_expirationTime", BindingFlags.NonPublic | BindingFlags.Instance);
+                    expirationTime.SetValue(tokenObj, expiry);
+                }
+
+                byte[] tokenBytes = (byte[])tokenObj.GetType().GetProperty("AccessToken", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(tokenObj, null);
+
+                tokenHash = GetTokenHash(tokenBytes);
+
+                return expiry;
+            }
+            catch (Exception)
+            {
+                tokenHash = "";
+                return null;
+            }
+        }
+
+        private string GetTokenHash(byte[] tokenBytes)
```

Comment:
This hash function duplicates the internal implementation and needs to stay in sync with it. Alternatively, you may consider following the reflection approach for it as well to ensure consistency.

Reply:
The existing GetTokenHash method from another helper class is not being used via reflection.